File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteAnnotationUpdater.java
package com.mapbox.services.android.navigation.v5.navigation;

import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.api.directions.v5.models.LegAnnotation;
import com.mapbox.api.directions.v5.models.RouteLeg;

import java.util.ArrayList;
import java.util.List;

class RouteAnnotationUpdater {

  DirectionsRoute update(DirectionsRoute oldRoute, DirectionsRoute annotationHolder,
                         int currentLegIndex) {

    List<RouteLeg> legs = new ArrayList<>(oldRoute.legs());

    for (int i = currentLegIndex; i < legs.size(); i++) {
      LegAnnotation updatedAnnotation = annotationHolder.legs().get(i - currentLegIndex).annotation();
      legs.set(i, legs.get(i).toBuilder().annotation(updatedAnnotation).build());
    }

    return oldRoute.toBuilder()
      .legs(legs)
      .build();
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteAnnotationUpdater.kt
package com.mapbox.services.android.navigation.v5.navigation

import com.mapbox.api.directions.v5.models.DirectionsRoute
import java.util.ArrayList

class RouteAnnotationUpdater {

    fun update(
        oldRoute: DirectionsRoute,
        annotationHolder: DirectionsRoute?,
        currentLegIndex: Int
    ): DirectionsRoute {
        oldRoute.legs()?.let { routeLegList ->
            val legs = ArrayList(routeLegList)
            for (i in currentLegIndex until legs.size) {
                annotationHolder?.legs().let { routeLegs ->
                    routeLegs?.let { routeLegs ->
                        val updatedAnnotation = routeLegs[i - currentLegIndex].annotation()
                        legs[i] = legs[i].toBuilder().annotation(updatedAnnotation).build()
                    }
                }
            }
            return oldRoute.toBuilder()
                    .legs(legs)
                    .build()
        }
        return oldRoute
    }
}
